cmd/skopeo/utils.go

```diff
@@ -257,3 +279,74 @@ func adjustUsage(c *cobra.Command) {
 	c.SetUsageTemplate(usageTemplate)
 	c.DisableFlagsInUseLine = true
 }
+
+func isRetryable(err error) bool {
+	err = errors.Cause(err)
+
+	if err == context.Canceled || err == context.DeadlineExceeded {
+		return false
+	}
+
+	type unwrapper interface {
+		Unwrap() error
+	}
+
+	switch e := err.(type) {
+
+	case unwrapper:
+		err = e.Unwrap()
+		return isRetryable(err)
+	case errcode.Error:
+		switch e.Code {
+		case errcode.ErrorCodeUnauthorized, errcodev2.ErrorCodeNameUnknown, errcodev2.ErrorCodeManifestUnknown:
+			return false
+		}
+		return true
+	case *net.OpError:
+		return isRetryable(e.Err)
+	case *url.Error:
+		return isRetryable(e.Err)
+	case syscall.Errno:
+		return !(e == syscall.ECONNREFUSED)
+	case errcode.Errors:
+		// if this error is a group of errors, process them all in turn
+		for i := range e {
+			if !isRetryable(e[i]) {
+				return false
+			}
+		}
+		return true
+	case *multierror.Error:
+		// if this error is a group of errors, process them all in turn
+		for i := range e.Errors {
+			if !isRetryable(e.Errors[i]) {
+				return false
+			}
+		}
+		return true
+	}
+
+	return true
+}
+
+func retryIfNecessary(ctx context.Context, operation func() error, retryOptions *retryOptions, commandTimeout time.Duration) error {
+	err := operation()
+	for attempt := 0; err != nil && isRetryable(err) && attempt < retryOptions.maxRetry; attempt++ {
+		logrus.Infof("Warning: failed, retrying in %s ... (%d/%d)", exponentialDelayTime(attempt, retryOptions.retryDelay), attempt+1, retryOptions.maxRetry)
```

Comment:
(Eventually it might be useful to log the error here — or log it only if it is different from the error on the previous attempt. But that would require somehow dealing with the very long errors we report so that it doesn’t overwhelm the “waiting now” information, and solving that can easily wait for a future PR.)

Reply:
ok, I can solve that in future PR. Would this make `isRetryable(err)` return both bool and an error string, so inside retryIfNecessary the error will be compared?

Reply:
This function can just collect the `err` value from the previous loop iteration. I wouldn’t typically expect the most-nested error used in `isRetryable` to actually make the decision to stay the same while the outermost error changes — and if that _were_ the case, that would be a good reason to report the different error (e.g. before a proxy was failing, then the registry was failing, then the underlying storage of a registry was failing, or something like that).